### PR TITLE
Fix FEInterface for MG support

### PR DESCRIPTION
**Pull request by mwichro**

### CODE REVIEW

File: include/deal.II/fe/fe_interface_values.h [hunk 1]
```diff
@@ -694,6 +724,38 @@ FEInterfaceValues<dim, spacedim>::at_boundary() const
 template <int dim, int spacedim>
 std::vector<types::global_dof_index>
 FEInterfaceValues<dim, spacedim>::get_interface_dof_indices() const
+{
+#  ifdef DEBUG
+  Assert(
+    is_level_cell == false,
+    ExcMessage(
+      "get_dof_indices() only works on active cells. Use get_mg_interface_dof_indices instead."))
```
Comment (masterleinad) [inline include/deal.II/fe/fe_interface_values.h:732]:
```suggestion
      "get_interface_dof_indices() only works on active cells. Use get_mg_interface_dof_indices instead."))
```

File: include/deal.II/fe/fe_interface_values.h [hunk 2]
```diff
@@ -694,6 +724,38 @@ FEInterfaceValues<dim, spacedim>::at_boundary() const
...
+
+template <int dim, int spacedim>
+std::vector<types::global_dof_index>
+FEInterfaceValues<dim, spacedim>::get_mg_interface_dof_indices() const
+{
+#  ifdef DEBUG
+  Assert(
+    is_level_cell == true,
+    ExcMessage(
+      "get_mg_dof_indices() only works on mg cells. Use get_interface_dof_indices instead."))
```
Comment (masterleinad) [inline include/deal.II/fe/fe_interface_values.h:748]:
```suggestion
      "get_mg_interface_dof_indices() only works on mg cells. Use get_interface_dof_indices instead."))
```

File: include/deal.II/fe/fe_interface_values.h [hunk 3]
```diff
@@ -694,6 +724,38 @@ FEInterfaceValues<dim, spacedim>::at_boundary() const
...
+#  endif
+
+    return interface_dof_indices;
+}
+
+
+
+template <int dim, int spacedim>
+std::vector<types::global_dof_index>
+FEInterfaceValues<dim, spacedim>::get_active_or_mg_dof_indices() const
```
Comment (masterleinad) [inline include/deal.II/fe/fe_interface_values.h:758]:
```suggestion
FEInterfaceValues<dim, spacedim>::get_active_or_mg_interface_dof_indices() const
```

File: include/deal.II/fe/fe_interface_values.h [hunk 4]
```diff
@@ -235,6 +235,18 @@ class FEInterfaceValues
...
   get_interface_dof_indices() const;
 
+  /**
+   * Same as above but for level cells.
+   */
+  std::vector<types::global_dof_index>
+  get_mg_interface_dof_indices() const;
+
+  /**
+   * Get dof indices for generic loops.
```
Comment (bangerth) [inline include/deal.II/fe/fe_interface_values.h:245]: Can you expand what "generic loops" are? It's not a term we commonly use in deal.II (and in fact, I don't know what it means :-) )

File: include/deal.II/fe/fe_interface_values.h [hunk 5]
```diff
@@ -450,6 +462,10 @@ class FEInterfaceValues
    * as determined in reinit().
    */
   FEFaceValuesBase<dim> *fe_face_values_neighbor;
+
+#ifdef DEBUG
+  bool is_level_cell;
+#endif
```
Comment (bangerth) [inline include/deal.II/fe/fe_interface_values.h:468]: Let's not do this. We've tried to keep the debug and release libraries ABI compatible -- they may contain different code, but the sizes of structures and interfaces of functions should be the same. So just add the variable if you need it.

**Comment by mwichro** [comment]
@tjhei  (and @masterleinad )
1) I have no idea how to write quick test for that. I am able to get MG  preconditioner and covergence rate seems to be constant. I've come up with the idea of smoother today, so it's a fresh thing. It also works in parallel: MPI, distributed triangulation.
2) I was trying to get the same interface as DoFCellAccessor where all functions :
get_active_or_mg_dof_indices 
get_dof_indices 
get_mg_dof_indices 
coexist. This also allows some assertions, so it may be easier to locate the bugs.

@bangerth  I'll update documentation and remove #ifdef DEBUG

**Comment by masterleinad** [comment]
>     1. I have no idea how to write quick test for that.

Have a look at `tests/feinterface/fe_interface_values_01.cc`. Copy and modify accordingly.

**Comment by tjhei** [comment]
> 2\. as trying to get the same interface as DoFCellAccessor where all functions :

My vote would be to reuse ``get_interface_dof_indices``.

**Comment by mwichro** [comment]
@tjhei  I've changed to _get_dof_indices_ since this was the that returned active dof indices previously. Now in case of level cell it returns mg dof indices.
 I've fixed my name and email, but I don't know how to fix it in the commit.

**Comment by masterleinad** [comment]
The error message says
```
Running indentation test on Pull Request #9978
invalid author 'Aazaram' without firstname and lastname
hint: for possible solutions, consult the webpage:
      https://github.com/dealii/dealii/wiki/Indentation#commit-authorship
```
and on that page is another link describing how to change your username or email for older commits.

File: include/deal.II/fe/fe_interface_values.h [hunk 6]
```diff
@@ -450,6 +453,7 @@ class FEInterfaceValues
    * as determined in reinit().
    */
   FEFaceValuesBase<dim> *fe_face_values_neighbor;
+
```
Comment (tjhei) [inline include/deal.II/fe/fe_interface_values.h:456]: remove this line please

File: include/deal.II/fe/fe_interface_values.h [hunk 7]
```diff
@@ -228,6 +228,9 @@ class FEInterfaceValues
 
   /**
    * Return the set of joint DoF indices. This includes indices from both cells.
+   * If reinit was called on active cell the set of active dof indices are returned,
```
Comment (tjhei) [inline include/deal.II/fe/fe_interface_values.h:231]:
```suggestion
   * If reinit was called with an active cell iterator, the indices are based on the active indices (returned by `DoFCellAccessor::get_dof_indices()` ),
```

File: include/deal.II/fe/fe_interface_values.h [hunk 8]
```diff
@@ -228,6 +228,9 @@ class FEInterfaceValues
...
+   * in case of level cell (that is, if is_level_cell() return true )
```
Comment (tjhei) [inline include/deal.II/fe/fe_interface_values.h:232]: please update, this function is no longer here.

File: include/deal.II/fe/fe_interface_values.h [hunk 9]
```diff
@@ -542,10 +546,12 @@ FEInterfaceValues<dim, spacedim>::reinit(
...
     std::vector<types::global_dof_index> v(
       fe_face_values->get_fe().n_dofs_per_cell());
-    cell->get_dof_indices(v);
+    cell->get_active_or_mg_dof_indices(v);
     std::vector<types::global_dof_index> v2(
       fe_face_values_neighbor->get_fe().n_dofs_per_cell());
-    cell_neighbor->get_dof_indices(v2);
+    cell_neighbor->get_active_or_mg_dof_indices(v2);
+
+
```
Comment (tjhei) [inline include/deal.II/fe/fe_interface_values.h:554]: remove empty lines

File: include/deal.II/fe/fe_interface_values.h [hunk 10]
```diff
@@ -695,7 +701,8 @@ template <int dim, int spacedim>
 std::vector<types::global_dof_index>
 FEInterfaceValues<dim, spacedim>::get_interface_dof_indices() const
 {
-  return interface_dof_indices;
+
+    return interface_dof_indices;
```
Comment (tjhei) [inline include/deal.II/fe/fe_interface_values.h:705]: please revert this change as well.

**Comment by tjhei** [comment]
Thank you, that looks much better now! Would you mind squashing your commits into a single commit? I am happy to contribute a test-case so you can see how to do it.

**Comment by luca-heltai** [comment]
Somehow tidy is complaining about something unrelated to this PR.

**Comment by masterleinad** [comment]
> Somehow tidy is complaining about something unrelated to this PR.

Yes, we need to rebase to get rid of it. I am fine with ignoring it here.
